### PR TITLE
gh-280: add `mypy` to codebase

### PR DESCRIPTION
**Pull request by paddyroddy**
Adding `mypy` to a pre-existing codebase is never easy. I initially attempted this in #308, but have since split this up into separate issues:
- [x] #347
- [x] #356
- [ ] #358

In this PR, `mypy` is added but in order for CI to pass, the `# type: ignore[]` syntax is used throughout. I didn't want to tackle these here too (see #308) as it gets quite messy.

One thing I have done (following #356) is change every empty `npt.NDArray` to `npt.NDArray[typing.Any]` (see #330), as this actually results in fewer errors than leaving them all blank. Ideally, we'd like to fill in as many of the `typing.Any` as possible - they're a bit useless by themselves. However, that is not the priority for now. Plus, I expect typing to change when #67 is tackled.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -92,6 +92,19 @@ build.targets.sdist.exclude = [
 ]
 version.source = "vcs"
 
+[tool.mypy]
```
Comment (paddyroddy) [inline pyproject.toml:95]: Some of these are related to #187

File: pyproject.toml [hunk 2]
```diff
@@ -92,6 +92,19 @@ build.targets.sdist.exclude = [
...
+    "ignore-without-code",
+    "redundant-expr",
+    "truthy-bool",
+]
+plugins = [
+    "numpy.typing.mypy_plugin",
+]
+strict = true
+warn_unreachable = true
+
```
Comment (paddyroddy) [inline pyproject.toml:107]: Some of these are related to #187

**Review by Saransh-cpp — APPROVED** [review]
This looks like a good start, thanks @paddyroddy!

**Comment by Saransh-cpp** [comment]
Ah 😬 I meant a good start for the other PR(s) that will add more type informations 😉

**Pull request merged**